File: algorithms/cracking_the_coding_interview/hash_tables_ransom_note.php
<?php
declare(strict_types=1);

const YES = 'Yes';
const NO = 'No';

/**
 * @param string[] $magazineNoteWords
 * @param string[] $ransomNoteWords
 * @return string
 */
function createReplica(array $magazineNoteWords, array $ransomNoteWords): string
{
    if (count($magazineNoteWords) < count($ransomNoteWords)) {
        return NO;
    }

    $magazineNoteMap = [];
    foreach ($magazineNoteWords as $word) {
        $magazineNoteMap[$word] = ($magazineNoteMap[$word] ?? 0) + 1;
    }

    foreach ($ransomNoteWords as $word) {
        if (!isset($magazineNoteMap[$word]) || $magazineNoteMap[$word] === 0) {
            return NO;
        }
        $magazineNoteMap[$word]--;
    }

    return YES;
}

$handle = fopen('php://stdin', 'r');
fscanf($handle, '%d %d', $m, $n);

$magazineNoteString = fgets($handle);
$magazineNoteWords = explode(' ', trim($magazineNoteString));
$ransomNoteString = fgets($handle);
$ransomNoteWords = explode(' ', trim($ransomNoteString));

echo createReplica($magazineNoteWords, $ransomNoteWords);

File: algorithms/cracking_the_coding_interview/hash_tables_ransom_note.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const (
	YES = "Yes"
	NO  = "No"
)

func createReplica(magazineNoteWords, ransomNoteWords []string) string {
	if len(magazineNoteWords) < len(ransomNoteWords) {
		return NO
	}

	magazineNoteMap := map[string]int{}
	for _, word := range magazineNoteWords {
		magazineNoteMap[word]++
	}

	for _, word := range ransomNoteWords {
		if value, ok := magazineNoteMap[word]; !ok || value == 0 {
			return NO
		}
		magazineNoteMap[word]--
	}

	return YES
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	reader.ReadString('\n')

	magazineNoteString, _ := reader.ReadString('\n')
	magazineNoteWords := strings.Fields(magazineNoteString)
	ransomNoteString, _ := reader.ReadString('\n')
	ransomNoteWords := strings.Fields(ransomNoteString)

	fmt.Println(createReplica(magazineNoteWords, ransomNoteWords))
}
